ubiquity/dell-bootstrap.py

```diff
@@ -760,6 +761,16 @@ def prepare(self, unfiltered=False):
             self.log("rec_type %s, stage %d, device %s" % (rec_type, self.stage, self.device))
             if (rec_type == 'factory' and self.stage == 2) or rec_type == 'hdd':
                 self.fixup_factory_devices(rec_part)
+            if rec_type == 'hdd':
+                # copy old mok key so that user don't need to enroll it again.
+                rootfs = mount[0:-1] + '3'
+                self.log("old rootfs from %s" % rootfs)
+                misc.execute_root('mount', '-o', 'ro', rootfs, '/mnt')
+                if os.path.exists('/mnt/var/lib/shim-signed/mok/MOK.priv') and os.path.exists('/mnt/var/lib/shim-signed/mok/MOK.der'):
+                    misc.execute_root('cp', '-f', '/mnt/var/lib/shim-signed/mok/MOK.der','/tmp')
+                    misc.execute_root('cp', '-f', '/mnt/var/lib/shim-signed/mok/MOK.priv','/tmp')
```

Comment:
This seems a bit fragile to me.

If the reinstall is interrupted after it's been started the file gets lost because it's sitting in `/tmp`.  I think it would be smarter to copy the file to temporary storage or memory and then immediately restore it on the next opportunity rather than as a late script step.

Reply:
I thought about that, it seems the only persistent storage here is the recovery partition, if the key is copied to recovery partition then user cancel the process. The key will be keeped in recovery partition which not what we expected.

here we keep key in /tmp, there're 3 scenarios in this stage.
1. the recovery is done completely - then it works well.

2. user cancel the recovery in the dialog page in the begin of the reboot - then the key will be lost as expected because nothing will be changed on the target system.

3. user cancel the recovery in the middle of recovery, force reboot by pressing the power button. Then the system will stay in non-stable status and get red grub. - in this case, the system is expected to full refreshed, so the key is expected to re-generated with no harm.

it seems all 3 scenarios are what we expected.

Reply:
I guess that's true.  OK.